src/python/pants/backend/python/rules/pex.py

```diff
@@ -67,13 +74,80 @@ def create_from_adaptors(
         return PexInterpreterConstraints(constraint_set=tuple(sorted(interpreter_constraints)))
 
 
+@dataclass(frozen=True)
+class PexRequirementConstraints:
+    """A collection of paths to constraint files and/or Requirement-style strings."""
+
+    constraint_file_paths: Tuple[str, ...] = ()
+    raw_constraints: Tuple[str, ...] = ()
+    generated_file_name: ClassVar[str] = "requirement_constraints.generated.txt"
+
+    @classmethod
+    def create_from_raw_constraints(cls, constraints: Iterable[str]) -> "PexRequirementConstraints":
+        """The constraints may be paths to constraint files (prefixed with `@`) and/or requirement
+        strings."""
+        constraint_file_paths = set()
+        raw_constraints = set()
+        for constraint in constraints:
+            if constraint.startswith("@"):
+                constraint_file_paths.add(constraint[1:])
+            else:
+                raw_constraints.add(constraint)
+        return PexRequirementConstraints(
+            constraint_file_paths=tuple(sorted(constraint_file_paths)),
+            raw_constraints=tuple(sorted(raw_constraints)),
+        )
+
+    @classmethod
+    def create_from_setup(cls, python_setup: PythonSetup) -> "PexRequirementConstraints":
+        return cls.create_from_raw_constraints(python_setup.requirement_constraints)
+
+    @classmethod
+    def create_from_adaptors(
+        cls, adaptors: Iterable[PythonTargetAdaptor], python_setup: PythonSetup
+    ) -> "PexRequirementConstraints":
+        merged_constraints = itertools.chain.from_iterable(
+            python_setup.resolve_requirement_constraints(getattr(adaptor, "constraints", None))
```

Comment:
Since we're accepting a graph of requirement constraints instead of just 1 from a `python_binary` (or two if the global default is defined) - is the merge order / semantics correct here? Presumably correct would mean closer to the root target wins and for same closeness declaration order in the BUILD file (or in imports? under buidgen).